kernel/src/vm/vmar/mod.rs

```diff
@@ -364,59 +364,41 @@ impl Vmar_ {
         Ok(())
     }
 
-    fn check_overwrite(&self, mapping_range: Range<usize>, can_overwrite: bool) -> Result<()> {
-        let inner = self.inner.read();
-
-        if !can_overwrite && inner.vm_mappings.find(&mapping_range).next().is_some() {
-            return_errno_with_message!(
-                Errno::EACCES,
-                "mapping range overlapped with another mapping"
-            );
-        }
-
-        Ok(())
-    }
-
     /// Returns the attached `VmSpace`.
     fn vm_space(&self) -> &Arc<VmSpace> {
         &self.vm_space
     }
 
-    /// Maps a `VmMapping` to this VMAR.
-    fn add_mapping(&self, mapping: VmMapping) {
-        self.inner.write().vm_mappings.insert(mapping);
-    }
-
-    fn allocate_free_region_for_mapping(
+    fn allocate_free_region_and_add_mapping(
```

Comment:
Good job! I think the changes in this PR are reasonable. Yet I find the current API semantics a bit strange. `allocate_free_region_and_add_mapping` actually means "allocating a free region for the input mapping and adding this mapping to the VMAR", which is overly complicated. 

I think one way is naming this method as `build_mapping` and let it accept `VmarMapOptions` as the parameter. Then we can remove `build_mapping` in `VmMapping` can call this methd in `VmarMapOptions.build()` directly.

The other way is making the `alloc_free_region_*` in `VmarInner` public to `vm_mapping` module and do the instructions in `VmMapping::build_mapping` (I think keep the name `build_mapping` is ok, just add sufficient documents), then we can remove this method in `Vmar`.